app/uk/gov/hmrc/helptosavefrontend/forms/BankDetailsValidation.scala

```diff
@@ -0,0 +1,172 @@
+
+    override def bind(key: String, data: Map[String, String]): Either[Seq[FormError], String] = {
+      val validation: ValidOrErrorStrings[String] =
+        data.get(key)
+          .map(_.cleanupSpecialCharacters.removeAllSpaces)
+          .fold(invalid[String](ErrorMessages.accountNumberIncorrectFormat)){ s ⇒
+            validatedFromBoolean(s)(s ⇒ s.length === accountNumberLength && s.forall(_.isDigit), ErrorMessages.accountNumberIncorrectFormat)
+          }
+
+      validation.toEither.leftMap(_.map(e ⇒ FormError(key, e)).toList)
+    }
+
+    override def unbind(key: String, value: String): Map[String, String] =
+      text.withPrefix(key).unbind(value)
+  }
+
+  val rollNumberFormatter: Formatter[Option[String]] = new Formatter[Option[String]] {
+
+    override def bind(key: String, data: Map[String, String]): Either[Seq[FormError], Option[String]] = {
+      val validation: ValidOrErrorStrings[Option[String]] =
+        data.get(key)
+          .map(_.cleanupSpecialCharacters.removeAllSpaces)
+          .filter(_.nonEmpty)
+          .fold[ValidOrErrorStrings[Option[String]]](Valid(None)){ s ⇒
+            if (s.length < rollNumberMinLength) {
+              invalid(ErrorMessages.rollNumberTooShort)
+            } else if (s.length > rollNumberMaxLength) {
+              invalid(ErrorMessages.rollNumberTooLong)
+            } else {
+              Valid(Some(s))
+            }
+          }
+
+      validation.toEither.leftMap(_.map(e ⇒ FormError(key, e)).toList)
+    }
+
+    override def unbind(key: String, value: Option[String]): Map[String, String] =
+      optional(text).withPrefix(key).unbind(value)
+  }
+
+  val accountNameFormatter: Formatter[String] = new Formatter[String] {
+
+    override def bind(key: String, data: Map[String, String]): Either[Seq[FormError], String] = {
+      val validation: ValidOrErrorStrings[String] =
+        data.get(key)
+          .map(_.cleanupSpecialCharacters.trim)
+          .fold(invalid[String](ErrorMessages.accountNameTooShort)){ s ⇒
+            if (s.length < accountNameMinLength) {
+              invalid(ErrorMessages.accountNameTooShort)
+            } else if (s.length > accountNameMaxLength) {
+              invalid(ErrorMessages.accountNameTooLong)
+            } else {
+              Valid(s)
+            }
+          }
+
+      validation.toEither.leftMap(_.map(e ⇒ FormError(key, e)).toList)
+    }
+
+    override def unbind(key: String, value: String): Map[String, String] =
+      text.withPrefix(key).unbind(value)
+  }
+
+}
+
+object BankDetailsValidation {
+
+  private[forms] object ErrorMessages {
+
+    val sortCodeIncorrectFormat = "sort_code_incorrect_format"
+
+    val accountNumberIncorrectFormat = "account_number_incorrect_format"
+
+    val rollNumberTooShort = "roll_number_too_short"
+
+    val rollNumberTooLong = "roll_number_too_long"
+
+    val accountNameTooShort = "account_name_too_short"
+
+    val accountNameTooLong = "account_name_too_long"
```

Comment:
also i was thinking these error key strings should appear the same way in messages.conf too, but i see in its `sort-code.incorrect-format` in the messages file, cant we keep the same key at both places ? 
keeping the same key at both places help in searching/finding for those occurrences in the whole codebase ?

Reply:
it doesn't really matter what the values of these strings are in reality - all they do is it facilitate the interface between the BE and the FE view. They're independent of what's in the messages file. The values themselves don't drive any logic except in the `FormOps` implicit class